### PR TITLE
fix: Apply reporting timezones for BigQuery functions (#25451)

### LINKED ISSUE #12380: BigQuery filter by "current day" (today) showing items from yesterday
**Describe the bug**
If I set my filter to to be "today" (Today is the 22nd), it will show results from yesterday:
image

You can see in the image above that its set for "today" (today is the 22nd) and that it is showing items from the 21st (yesterday).

If I were to filter on April 22nd, it shows the correct results:
image

**To Reproduce**
1. Create a new question
2. Use a  BigQuery database
3. filter by a column that has dates, filter by "current" "day"
4. See its displaying stuff from yesterday.

**Expected behavior**
I expect it to show only items with the current day (April 22nd) not items in the 24 hour period.

**Screenshots**

**Information about your Metabase Installation:**
You can get this information by going to Admin -> Troubleshooting.

- Your browser and the version: (e.x. Chrome 52.1, Firefox 48.0, Safari 11.1, …)
- Your operating system: Windows 10, Chrome
- Your databases: Big Query
- Metabase version:v 0.35.3
- Metabase hosting environment: Docker
- Metabase internal database: MySQL

**Severity**
This is annoying, causes users to not trust the data.

### LINKED ISSUE #25080: Incorrect last-X-weeks/months/years calculation on BigQuery
**Describe the bug**
In b9cedccc7f8, @camsaul mostly fixed the BigQuery driver so that it could express the date calculations needed to drive X-weeks/months/years-ago calculations. This gets used whenever we create a filter by last X weeks/month/years.

However, this does not work correctly when the reporting timezone is not UTC. We can see why by testing out the parts of the generated SQL. For a "last 2 weeks, grouped by week" question, with the reporting timezone set to 'Pacific/Auckland' and the first day of the week set to Monday, the SQL looks sometime like this:

```
SELECT timestamp_trunc(`bi.mytable`.`event_at`, week(monday)) AS `event_at`, sum(`bi.mytable`.`somefield`) AS `sum`
FROM `bi.mytable`
WHERE (`bi.mytable`.`event_at` >= CAST(datetime_trunc(datetime_add(current_datetime(), INTERVAL -2 week), week(monday)) AS timestamp)
   AND `bi.mytable`.`event_at` < CAST(datetime_trunc(current_datetime(), week(monday)) AS timestamp))
GROUP BY `event_at`
ORDER BY `event_at` ASC
```

Note that the reporting timezone doesn't appear anywhere in the query. Running the expressions in the BigQuery console:

```
SELECT CAST(datetime_trunc(datetime_add(current_datetime(), INTERVAL -2 week), week(monday)) AS timestamp)
```

Produces `2022-08-15 00:00:00 UTC` - so the week cutoff is for UTC, not the local timezone.

The first issue is that the grouping part, `timestamp_trunc(`bi.mytable`.`event_at`, week(monday))`, needs to specify the timezone - `timestamp_trunc(`bi.mytable`.`event_at`, week(monday), 'Pacific/Auckland')` in my case.

The second part, where the relative calculation is performed, is more complicated. Firstly we would need to specify the timezone to `current_datetime()`, ie. `current_datetime('Pacific/Auckland')`. Unfortunately this isn't enough - the timezone still gets dropped, and the timestamp ends up at 00:00:00 UTC rather than 00:00:00 Pacific/Auckland. (I think BigQuery's handling of datetime timezones is poor here.)

Given that `timestamp_trunc` works correctly (if we give it the timezone), one option is to move the cast so that we do only the absolute minimum in converted datetimes, which is the relative time calculation - ie. just the `datetime_add` - then cast, then use `timestamp_trunc` with reporting timezone. This looks to me to be much simpler than trying to fix the timezone on the `datetime_add`/`datetime_trunc`.

In this approach, we don't even need to specify the timezone to `current_datetime` - it's fine to use UTC for the current time and the relative date calculation, as if we then convert it to a timestamp it's UTC anyway, and only then do we run the truncation in the specific reporting timezone.

So my query above would become:

```
SELECT timestamp_trunc(`bi.mytable`.`event_at`, week(monday), 'Pacific/Auckland') AS `event_at`, sum(`bi.mytable`.`somefield`) AS `sum`
FROM `bi.mytable`
WHERE (`bi.mytable`.`event_at` >= timestamp_trunc(CAST(datetime_add(current_datetime(), INTERVAL -2 week) AS timestamp), week(monday), 'Pacific/Auckland')
   AND `bi.mytable`.`event_at` < timestamp_trunc(CAST(current_datetime() AS timestamp), week(monday), 'Pacific/Auckland'))
GROUP BY `event_at`
ORDER BY `event_at` ASC
```

**To Reproduce**
1. Setup up a BigQuery table with a timestamp column
2. Set the reporting timezone to something other than UTC
3. Create a new question summarising some value by the timestamp at week granularity
4. Filter to the last 2 weeks
5. Observe that rows are filtered and grouped based on the UTC week boundaries, not the reporting timezone

**Expected behavior**
Week/month/year grouping should be performed in the reporting timezone for BigQuery.

**Information about your Metabase Installation:**

You can get this information by going to Admin -> Troubleshooting.

Your browser and the version: Chrome 104
Your operating system: macOS 12.3.1
Your databases: BigQuery
Metabase version: v0.44.1
Metabase hosting environment: Docker
Metabase internal database: H2

**Severity**
How severe an issue is this bug to you? Is this annoying, blocking some users, blocking an upgrade or blocking your usage of Metabase entirely?
Blocking use for financials and causing some weird UX behavior.

### PR DESCRIPTION
**Pull request by nemanjaglumac**
#25451 opened from within Metabase organization in order to run against all CI checks.

Fixes https://github.com/metabase/metabase/issues/25080
Fixes https://github.com/metabase/metabase/issues/12380
Implements https://github.com/metabase/metabase/issues/10772

Huge thanks and all the credit goes to @willbryant!

For more details and a full description, please see #25451

### CODE REVIEW

File: modules/drivers/bigquery-cloud-sdk/test/metabase/test/data/bigquery_cloud_sdk.clj [hunk 1]
```diff
@@ -116,7 +116,7 @@
 
 ;; Fields must contain only letters, numbers, and underscores, start with a letter or underscore, and be at most 128
 ;; characters long.
-(def ^:private ValidFieldName #"^[A-Za-z_]\w{0,127}$")
+(def ^:private ValidFieldName #"^[A-Za-z_][\w-]{0,127}$")
```
Comment (bshepherdson) [inline modules/drivers/bigquery-cloud-sdk/test/metabase/test/data/bigquery_cloud_sdk.clj:119]: This doesn't match the description above - I suppose dashes are actually legal in BigQuery?

**Review by bshepherdson — APPROVED** [review]

File: modules/drivers/bigquery-cloud-sdk/test/metabase/test/data/bigquery_cloud_sdk.clj (hunk 1 above)
Comment (willbryant) [reply modules/drivers/bigquery-cloud-sdk/test/metabase/test/data/bigquery_cloud_sdk.clj:119]: Thanks, good spotting - meant to revert that after proving it wasn't accepted by BigQuery.

Fixed in the origin pull request in 10e13647b1.

Comment (nemanjaglumac) [reply modules/drivers/bigquery-cloud-sdk/test/metabase/test/data/bigquery_cloud_sdk.clj]: Pulled from the upstream and updated this branch.

**Comment by willbryant** [comment]
Are we waiting on anything now? Keen to get this merged.

**Comment by nemanjaglumac** [comment]
@willbryant the CI is still failing. I'll take a look into that.

@camsaul 's requested changes are still blocking THIS pr for whatever reason, although he requested changes in your own PR. Cam, maybe you can approve the changes there?

**Comment by nemanjaglumac** [comment]
Rebased on the latest `master` because the PR became stale.

**Comment by tsmacdonald** [comment]
Thanks @willbryant !

**Comment by willbryant** [comment]
Awesome, thanks @nemanjaglumac :)

**Pull request merged**